### DISCUSSION TITLE
Sharing SPI bus with EVE

### DISCUSSION THREAD

**Discussion by dominicdumont** [RudolphRiedel/FT800-FT813#44] [Q&A]
Hi Rudolph,
  I am using your great library on a couple of PIC32 projects and it works really well. I now need to use an EVE display with an ESP32. I was able to easily make it work, but I am having some trouble sharing the SPI bus with other devices.

In my PIC32 projects, I do not use any library apart from yours (I use **_c_** to directly talk to the SPI device) - so I simply select / deselect (CS) the device on the SPI bus and all works well.

But for the ESP32, I am using Arduino, so I end up with a bunch of libraries that I don't always know exactly how they are using the SPI bus. For this project, I am trying to share a [MAX3421E](https://www.maximintegrated.com/en/products/interface/controllers-expanders/MAX3421E.html) (USB host chip through SPI) with your EVE library. I am using [this ](https://github.com/felis/USB_Host_Shield_2.0) library as the interface.

I can successfully use both device, but not at the same time. If I only use EVE in my code, all is good. If I only use MAX3421E in my code, all is good. But if I try to use EVE, then try to use MAX3421E, EVE will stop to respond to command after I start using MAX3421E (and vice-versa). So I was wondering if you're EVE library requires _exclusive_ use of the SPI port?

(On another board  I made with an ESP32 (no EVE display) I am able to share the same SPI bus with an RTC and FRAM.)

I know there might be missing some details (which I can provide if needed) but if the library requires exclusive use of the SPI port, all other details will be irrelevant :)

Since the EPS32 has 2 SPI ports, I could just use 1 port for EVE and 1 port to share with other device - but the current board I made have both on the same port so taking a chance to ask :)

Thanks for the great library!

Dominic

**Comment by RudolphRiedel** [comment]
I am afraid that this might be a tough one.
The MAX3421 library is using the SPI class from arduino-esp32.
Since the SPI class does not support DMA I ended up using the underlying ESP-IDF instead.

Looking at the sources, there might be a way.
In normal mode my driver is claiming the SPI before chip-select and releasing it afterwards.
I implemented this after experimenting with the ESP32 and finding it to work a little bit faster this way.
In DMA mode it does not claim and release the SPI.

Try commenting out the  #define EVE_DMA line in EVE_target.h, this is line 1869 for the ESP32 target.

If this works we might get it to work with DMA as well by adding two extra lines to EVE_cpp_target.cpp.

Well, this assumes that the SPI class plays nice and just waits as long as the SPI is claimed.
Likewise claiming the SPI should not return untill the SPI was free.

**Reply by dominicdumont** [reply]
Hi Rudolph,
  Thanks for the reply. I tried but unfortunately the same issue persist.  I will use the other SPI port for EVE. Thanks for looking into it.

**Reply by RudolphRiedel** [reply]
As high-level as SPI is handled for the ESP32, this should be possible.
But then it very well could be that the ESP32-Arduino class and the ESP-IDF can not be used in parallel.

**Comment by RudolphRiedel** [comment]
Oh, forgot, since I do not currently have a PIC32 target, would you like to share what you have so I can add it to EVE_target.h?

**Reply by dominicdumont** [reply]
I don't really use your library "as-is" for my PIC32 project. I originally had written code similar as yours to drive my EVE displays - but let just say that mine was not as complete as yours :) So when I found your library, I just took the parts I needed and replace my code with yours. Next PIC32 project I do, Ill try to integrate your library "as-is" so settings can be shared.

**Reply by RudolphRiedel** [reply]
Especially when you already know the controller, adding support to my library is no big deal.
PIC32 is a difficult topic for me, I am really not a fan of MPLAB-X to put it friendly.
But then, so far no one asked for PIC support.

**Reply by dominicdumont** [reply]
I am not a fan either of MPLAB-X - particularly the bloated Harmony framework. But I do like the PIC32 :) I had ok success with MIKROE [compilers](https://www.mikroe.com/necto) so I can stay away from MPLAB-X.

**Reply by RudolphRiedel** [reply]
I would prefer a VSCode / PlatformIO solution but I just had a look and the "Micochip PIC32" platform only supports Arduino as framework.

And my impression of Harmony is very different.
This for example:
https://github.com/Microchip-MPLAB-Harmony/csp_apps_sam_c20_c21
https://github.com/Microchip-MPLAB-Harmony/csp_apps_sam_c20_c21/blob/master/apps/sercom/spi/spi_eeprom_write_read/firmware/src/config/sam_c21n_xpro/peripheral/sercom/spi_master/plib_sercom5_spi_master.c

I think this is rather clean and most importantly one could learn from it.
Bloat is what Atmel put out as ASF.
Ok, that is already generated code, MPLAB-X by itself and the Harmony Configurator is crazy large.
If only one could decide for which families it installs...

Unfortunately Microchip decided to make all the ATSAM includes completely incompatible with the Atmel Studio includes.
So I can't just copy snippets from their examples, they would not compile.

**Reply by dominicdumont** [reply]
I think I must have tried MPLAB-X + harmony at least 4-5 times in the past years. I find that the learning curve is enormous. 

3 months ago I had no interest in the Arduino platform, but it's everywhere so I decided to take a look. I started reading tutorials then actually tried it. Now I can run an ESP32 with Wifi (MQTT), USB and EVE, using VS Code/Platform IO. I was able to do all that part time (week-ends / free time) under 1 month. 

Using Harmony, I highly doubt I could do the equivalent **as fast**. So I am having a hard time justifying investing time learning it. So now for critical stuff, I still use a PIC32. But for other stuff I consider less critical or less time constrained for my particular projects (Wifi, USB, etc.), I use an ESP32 running the Arduino platform as my "front-end" and use a PIC32 as my "back-end" to do all the critical stuff.
